Mathlib/Topology/Order/HullKernel.lean

```diff
@@ -0,0 +1,289 @@
+/-
+Copyright (c) 2024 Christopher Hoskin. All rights reserved
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import Mathlib.Topology.Order.LowerUpperTopology
+import Mathlib.Order.Irreducible
+import Mathlib.Data.Set.Subset
+import Mathlib.Order.Interval.Set.Monotone
+import Mathlib.Topology.Sets.Closeds
+
+/-!
+# Hull-Kernel Topology
+
+Let `α` be a `CompleteLattice` and let `T` be a subset of `α`. The relative `Topology.lower` on `T`
+takes on a particularly simple form when every element of `T` is `InfPrime` in `α`. In this case,
+the relative-open sets are exactly the sets of the form `T ↓∩ (Ici a)ᶜ` for some `a` in `α`.
+
+The pair of maps `S → ⊓ S` and `a → T ↓∩ (Ici a)` are often referred to as the kernel and the hull.
+They form an antitone Galois connection between the subsets of `T` and `α`. When `α` can be
+generated from `T` by taking infs, this becomes a Galois insertion and the topological closure
+coincides with the closure arising from the Galois insertion. For this reason the relative lower
+topology on `T` is often refereed to as the hull-kernel topology. The names Jacobson topology and
+structure topology also occur in the literature.
+
+## Main statements
+
+- `PrimitiveSpectrum.relativeLowerIsTopologicalBasis` - the sets `T ↓∩ (Ici a)` form a basis for the
+  relative lower topology on `T`.
+- `PrimitiveSpectrum.isOpen_iff` - for a complete lattice, the sets `T ↓∩ (Ici a)` are the relative
+  topology.
+- `PrimitiveSpectrum.gc` - the kernel and the hull form a Galois connection
+- `PrimitiveSpectrum.gi` - when `T` generates `α`, the Galois connection becomes an insertion.
+- `PrimitiveSpectrum.lowerTopology_closureOperator_eq_gc_closureOperator` - relative lower topology
+  closure coincides with the "hull-kernel" closure arising from the Galois insertion.
+
+## Implementation notes
+
+The antitone Galois connection from `Set T` to `α` is implemented as a monitone Galois connection
+between `Set T` to `αᵒᵈ`.
+
+## Motivation
+
+The motivating example for the study of a set `T` of prime elements which generate `α` is the
+primitive spectrum of the lattice of M-ideals of a Banach space.
+
+## References
+
+* [Gierz et al, *A Compendium of Continuous Lattices*][GierzEtAl1980]
+
+## Tags
+
+lower topology, hull-kernel topology, Jacobson topology, structure topology, primitive spectrum
+
+-/
+
+variable {α}
+
+open TopologicalSpace
+open Topology
+open Set
+open Set.Notation
+
+section SemilatticeInf
+
+variable [SemilatticeInf α] [TopologicalSpace α] [IsLower α]
+
+variable {T : Set α} (hT : ∀ p ∈ T, InfPrime p)
```

Comment:
This is suspicious... I suspect I need to read the references to understand, but I won't have time before another week.

Reply:
I probably got the idea for this approach from Henriksen et al https://link.springer.com/article/10.1007/s000120050056. The hull kernel topology on the Primitive Spectrum of a C*-algebra is discussed in most introductory texts on C*-algebras. Various related concepts are used in more specialist applications.

However, one can get a long way with both the basic results and the specialist applications without reference to C*-algebras. Much of the theory depends just on the Primitive Spectrum being a dense order-generating subset of the prime elements of the lattice of norm-closed two-sided ideals of the C*-algebra. The advantage of taking this point of view is that the results then apply in more general situations like the lattice norm-closed ideals of a JB*-algebra or even the lattice of M-ideals of a Banach space. Henriksen et al also make a connection to minimal primes in algebraic lattices.

AFAIK we don't have two sided ideals in Mathlib yet, so I couldn't develop this theory specifically for C*-algebras even if I wanted to.

I don't know if there's an accepted term for a dense, order-generating subset of the prime elements of a lattice. I ended up calling such sets "primitive" in my doctoral thesis, but I was never very satisfied with that as a name.

Reply:
> I probably got the idea for this approach from Henriksen et al https://link.springer.com/article/10.1007/s000120050056. The hull kernel topology on the Primitive Spectrum of a C*-algebra is discussed in most introductory texts on C*-algebras. Various related concepts are used in more specialist applications.

> However, one can get a long way with both the basic results and the specialist applications without reference to C*-algebras. Much of the theory depends just on the Primitive Spectrum being a dense order-generating subset of the prime elements of the lattice of norm-closed two-sided ideals of the C*-algebra.

Maybe @j-loreaux will have an opinion on how your general machinery should interface with C*-algebras. What `T` will you take then? Shouldn't `ht : ∀ p ∈ T, InfPrime p` be automatically provided by typeclass inference then (by making that predicate a typeclass)?

> AFAIK we don't have two sided ideals in Mathlib yet, so I couldn't develop this theory specifically for C*-algebras even if I wanted to.

#13401 and #13902 are relevant.

Reply:
Let `A` be a Banach space and let `δₑA^*₁` be the set of extreme points of the dual unit ball. For each `x` in `δₑA^*₁` let `kₙ(ker x)` be the largest M-ideal of `A` contained in the kernel `ker x` of `x`. Then α would be the lattice of M-ideals in `A` and `T` would be `Prim A`, the set of M-ideals `{ kₙ(ker x) : x ∈ δₑA^*₁ }`.

Reply:
So `ht` here is a fact that you establiush once and for all, right? I think it should therefore be a typeclass